### PR TITLE
Bug 1285657 - Rewrite MXR links in comments to DXR

### PR DESCRIPTION
**Pull request by SebastinSanty**
[Bug 1285657 - Rewrite MXR links in comments to DXR](https://bugzilla.mozilla.org/show_bug.cgi?id=1285657)

Desc: See [Bug 1285657](https://bugzilla.mozilla.org/show_bug.cgi?id=1285657)

### CODE REVIEW

File: extensions/BMO/Extension.pm [hunk 1]
```diff
@@ -51,6 +51,9 @@ use List::Util qw(first);
 use Scalar::Util qw(blessed);
 use Sys::Syslog qw(:DEFAULT);
 use Text::Balanced qw( extract_bracketed extract_multiple );
+use URI;
+use URI::QueryParam;
+use URI::Split qw(uri_split uri_join);
```
Comment (dylanwh) [inline extensions/BMO/Extension.pm]: you needn't use the URI::Split, the URI object has methods available (I'll explain below)

File: extensions/BMO/Extension.pm [hunk 2]
```diff
@@ -728,6 +731,23 @@ sub bug_format_comment {
...
+            my ($scheme, $auth, $path, $query, $frag);
+            ($scheme, $auth, $path, $query, $frag) = uri_split($uri);
+            my $mark = $uri->query_param_delete("mark");
+            my $rev = $uri->query_param_delete("rev");
+            my $rev = $rev? "?rev=".$rev : '';
+            my $mark = $mark? $mark."," : '';
+            my $hash = $mark || $frag? "#": '';
+            return qq{<a href="$scheme://dxr.mozilla.org$path$rev$hash$mark$frag">$scheme://dxr.mozilla.org$path$rev$hash$mark$frag</a>};
+        }
+    });
```
Comment (dylanwh) [inline extensions/BMO/Extension.pm]: Something along the lines of...

````perl
            my $args  = shift;
            my $match = $args->{matches}->[0];
            my $uri   = URI->new($match);
            my $text  = html_quote($match);
            my $mark  = $uri->query_param_delete("mark");
            my $rev   = $uri->query_param_delete("rev");
            my (undef, $repo, $source, @path) = split( /\//, $uri->path );

            $uri->host("dxr.mozilla.org");

            if ($source ne 'source') {
                return qq{<a href="$text">$text</a>};
            }

            if ($mark) {
                $uri->fragment($mark);
            }

            if ($rev) {
                $uri->path("/$repo/rev/$rev/" . join('/', @path));
            }
            else {
                $uri->path("/$repo/source/" . join('/', @path));
            }

            my $uri_str = html_quote("$uri");
            return qq{<a href="$uri_str">$text [using dxr]</a>};
````

Comment (dylanwh) [inline extensions/BMO/Extension.pm]: note that code isn't complete:

something else is already in $uri->fragment, we should prepend that if $mark doesn't already contain it. Also most mxr links repeat the line number for single lines, which looks bad: 1234-1234 should replaced by 1234.

Comment (SebastinSanty) [inline extensions/BMO/Extension.pm]: I think is line numbers like 1234-1234 is automatically replaced by 1234.

**Comment by SebastinSanty** [comment]
Dylan, the patch has been fixed for changes mentioned by you. Can you please review it?

**Comment by floatingatoll** [comment]
@dylanwh Do you want this rebased forward still?

**Pull request closed without merging**